lib/runtime/wasmer/instance.go

```diff
@@ -94,6 +97,22 @@ func NewInstance(code []byte, cfg *Config) (*Instance, error) {
 		return nil, errors.New("code is empty")
 	}
 
+	// Substrate Wasm compression
+	// https://github.com/paritytech/substrate/blob/master/primitives/maybe-compressed-blob/src/lib.rs
+	const compressionFlag = []byte{82, 188, 83, 118, 70, 219, 142, 5}
+	if bytes.HasPrefix(code, compressionFlag) {
+		logger.Debug("Decompressing Wasm runtime code")
+		decoder, err := zstd.NewReader(nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create zstd decoder: %w", err)
+		}
+
+		code, err = decoder.DecodeAll(code[len(compressionFlag):], nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decompress compressed wasm code: %w", err)
+		}
```

Comment:
nit to reduce memory allocations so it's :electron:  faster

```suggestion
		decompressedCode := make([]byte, 0, len(code))
		_, err = decoder.DecodeAll(code[len(compressionFlag):], decompressedCode)
		if err != nil {
			return nil, fmt.Errorf("failed to decompress compressed wasm code: %w", err)
		}
		code = decompressedCode
```